torch/distributed/_sharded_tensor/api.py

```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
         # Barrier for all RPCs to finish on all ranks.
         rpc.api._all_gather(None)
 
+    def gather(
+        self,
+        dst: int = 0,
+        out: Optional[torch.Tensor] = None,
+    ) -> None:
+        """
+        Creates a full :class:`Tensor` on rank `dst` by gathering all sharded tensors.
+
+        The API needs to be called on all ranks in SPMD fashion. All ranks should have
+        the same `dst`. `out` should be a full size zero tensor on `dst` and
+        None on all other ranks.
+
+        Args:
+            dst(int): The rank where full tensor is constructed
+            out (:class `torch.Tensor` optional): The output full tensor. Needs to be provided
+            ONLY on `dst`
+        """
+        my_rank = dist.get_rank(self._process_group)
+        full_size = self.metadata().size
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
+
+        shard_tensors = self.local_shards()
+        world_size = dist.get_world_size(self._process_group)
+
+        gathered_shards = [None] * world_size
+        # all_gather_object is not efficient. will revise this part once
+        # NCCL support for dist.gather() is ready
+        with torch.cuda.device(curr_device):
+            dist.all_gather_object(
+                obj=shard_tensors,
+                object_list=gathered_shards,
+                group=self._process_group,
+            )
+
+        if my_rank == dst:
+            dims = len(full_size)
+            for shards in gathered_shards:
+                assert shards is not None, f"gathered shards cannot be None on {dst}"
+                for shard in shards:
+                    metadata = shard.metadata
+                    tensor = shard.tensor.to(curr_device)
```

Comment:
there might be case that the tensors gathered together can't fit into a single GPU device right (that's the main reason user might want to shard the tensor). If the whole point of this `gather` call is to gather the content and fill the passed in tensor, should we not care about the device and just copy the content to the `out`?

Reply:
I think it is a valid concern about size. @wanchaol so we can just do copy_ as you mentioned below. is it correct?